From: Jordan Bettcher <thebettcher@gmail.com>
Subject: v8: ppc: ensure ISA 2.07 (POWER8) compatibility across JIT codegen

Upstream V8 unconditionally emits several Power ISA 3.0 (POWER9) instructions:
1. `addpcis` (lnia) in `MacroAssembler::LoadPC`
2. `mcrxrx` in `MacroAssembler::MoveToCrFromXer`
3. `cnttzw` / `cnttzd` in `MacroAssembler::CountTrailingZerosU32/U64`
4. `modsw` / `modsd` / `moduw` / `modud` in Turbofan `code-generator-ppc.cc` and `MacroAssembler::Mod*`

On real POWER8 hardware or in POWER8 compatibility mode, these instructions
cause immediate `illegal instruction (4)` crashes.

This patch:
- Replaces `addpcis` with a `bcl`/`mflr`/`addi` sequence and adjusts `kPcLoadDelta` when not compiling for POWER9.
- Dispatches `MoveToCrFromXer` dynamically: uses `mcrxrx` on POWER9+ and `mcrxr` on POWER8.
- Synthesizes `CountTrailingZeros` via `subfic/andc/cntlz` on POWER8.
- Routes Turbofan modulo opcodes through `MacroAssembler::Mod*` and implements `div/mul/sub` fallbacks on POWER8.

--- a/v8/src/codegen/ppc/constants-ppc.h
+++ b/v8/src/codegen/ppc/constants-ppc.h
@@ -3065,7 +3065,11 @@ enum BOfield {        // Bits 25-21
 
 constexpr uint8_t kInstrSize = 4;
 constexpr uint8_t kInstrSizeLog2 = 2;
+#if defined(_ARCH_PWR9)
 constexpr uint8_t kPcLoadDelta = 4;
+#else
+constexpr uint8_t kPcLoadDelta = 12;
+#endif
 
 class Instruction {
  public:
--- a/v8/src/codegen/ppc/macro-assembler-ppc.h
+++ b/v8/src/codegen/ppc/macro-assembler-ppc.h
@@ -987,7 +987,11 @@
   // Usage: call the appropriate arithmetic function and then call one of the
   // flow control functions with the corresponding label.
   void MoveToCrFromXer(CRegister cr) {
+    if (CpuFeatures::IsSupported(PPC_9_PLUS)) {
       mcrxrx(cr);
+    } else {
+      mcrxr(cr);
+    }
   }
 
   // Performs a truncating conversion of a floating point number as used by
--- a/v8/src/codegen/ppc/macro-assembler-ppc.cc
+++ b/v8/src/codegen/ppc/macro-assembler-ppc.cc
@@ -1186,7 +1186,13 @@
 }
 
 void MacroAssembler::LoadPC(Register dst) {
+#if defined(_ARCH_PWR9)
   addpcis(dst, Operand(0));
+#else
+  bc(4, BA, 0, SetLK);
+  mflr(dst);
+  addi(dst, dst, Operand(8));
+#endif
 }
 
 void MacroAssembler::ComputeCodeStartAddress(Register dst) {
@@ -2602,19 +2608,43 @@
 }
 
 void MacroAssembler::ModS64(Register dst, Register src, Register value) {
-  modsd(dst, src, value);
+  if (CpuFeatures::IsSupported(PPC_9_PLUS)) {
+    modsd(dst, src, value);
+  } else {
+    divd(r0, src, value);
+    mulld(r0, r0, value);
+    sub(dst, src, r0);
+  }
 }
 
 void MacroAssembler::ModU64(Register dst, Register src, Register value) {
-  modud(dst, src, value);
+  if (CpuFeatures::IsSupported(PPC_9_PLUS)) {
+    modud(dst, src, value);
+  } else {
+    divdu(r0, src, value);
+    mulld(r0, r0, value);
+    sub(dst, src, r0);
+  }
 }
 
 void MacroAssembler::ModS32(Register dst, Register src, Register value) {
-  modsw(dst, src, value);
+  if (CpuFeatures::IsSupported(PPC_9_PLUS)) {
+    modsw(dst, src, value);
+  } else {
+    divw(r0, src, value);
+    mullw(r0, r0, value);
+    sub(dst, src, r0);
+  }
   extsw(dst, dst);
 }
 void MacroAssembler::ModU32(Register dst, Register src, Register value) {
-  moduw(dst, src, value);
+  if (CpuFeatures::IsSupported(PPC_9_PLUS)) {
+    moduw(dst, src, value);
+  } else {
+    divwu(r0, src, value);
+    mullw(r0, r0, value);
+    sub(dst, src, r0);
+  }
   ZeroExtWord32(dst, dst);
 }
 
@@ -4931,12 +4961,26 @@
 
 void MacroAssembler::CountTrailingZerosU32(Register dst, Register src,
                                            RCBit r) {
+  if (CpuFeatures::IsSupported(PPC_9_PLUS)) {
     cnttzw(dst, src, r);
+  } else {
+    addi(r0, src, Operand(-1));
+    andc(dst, r0, src);
+    cntlzw(dst, dst, r);
+    subfic(dst, dst, Operand(32));
+  }
 }
 
 void MacroAssembler::CountTrailingZerosU64(Register dst, Register src,
                                            RCBit r) {
+  if (CpuFeatures::IsSupported(PPC_9_PLUS)) {
     cnttzd(dst, src, r);
+  } else {
+    addi(r0, src, Operand(-1));
+    andc(dst, r0, src);
+    cntlzd(dst, dst, r);
+    subfic(dst, dst, Operand(64));
+  }
 }
 
 void MacroAssembler::ClearByteU64(Register dst, int byte_idx) {
--- a/v8/src/compiler/backend/ppc/code-generator-ppc.cc
+++ b/v8/src/compiler/backend/ppc/code-generator-ppc.cc
@@ -1527,16 +1527,16 @@
       ASSEMBLE_FLOAT_BINOP_RC(fdiv, MiscField::decode(instr->opcode()));
       break;
     case kPPC_Mod32:
-      __ modsw(i.OutputRegister(), i.InputRegister(0), i.InputRegister(1));
+      __ ModS32(i.OutputRegister(), i.InputRegister(0), i.InputRegister(1));
       break;
     case kPPC_Mod64:
-      __ modsd(i.OutputRegister(), i.InputRegister(0), i.InputRegister(1));
+      __ ModS64(i.OutputRegister(), i.InputRegister(0), i.InputRegister(1));
       break;
     case kPPC_ModU32:
-      __ moduw(i.OutputRegister(), i.InputRegister(0), i.InputRegister(1));
+      __ ModU32(i.OutputRegister(), i.InputRegister(0), i.InputRegister(1));
       break;
     case kPPC_ModU64:
-      __ modud(i.OutputRegister(), i.InputRegister(0), i.InputRegister(1));
+      __ ModU64(i.OutputRegister(), i.InputRegister(0), i.InputRegister(1));
       break;
     case kPPC_ModDouble:
       // TODO(bmeurer): We should really get rid of this special instruction,
